flopy/discretization/modeltime.py

```diff
@@ -71,21 +226,373 @@ def totim(self):
+        elif isinstance(datetime_obj, str):
+            str_rep = ModelTime._get_datetime_string_format(datetime_obj)
+            datetime_obj = datetime.datetime.strptime(datetime_obj, str_rep)
+
+        else:
+            raise NotImplementedError(
+                f"{type(datetime_obj)} date representations "
+                f"are not currently supported"
+            )
+
+        return datetime_obj
+
+    def get_totim(self, kper, kstp=None):
+        """
+        Method to get the total simulation time at the end of a given
+        stress-period or stress-period and time-step combination
+
+        Parameters
+        ----------
+        kper : int
+            zero based stress-period number
+        kstp : int or None
+            optional zero based time-step number
+
+        Returns
+        -------
+            totim : float
+        """
+        if kstp is None:
+            kstp = self.nstp[kper] - 1
+
+        if not self._totim_dict:
+            self._set_totim_dict()
+
+        if (kper, kstp) not in self._totim_dict:
+            raise KeyError(
+                f"(kper, kstp): ({kper} {kstp}) not a valid combination of "
+                f"stress period and time step"
+            )
+
+        return self._totim_dict[(kper, kstp)]
+
+    def get_datetime(self, kper, kstp=None):
+        """
+        Method to get the datetime at the end of a given stress period or
+        stress period and time step combination
+
+        Parameters
+        ----------
+        kper : int
+            zero based modflow stress period number
+        kstp : int
+            zero based timestep number
+
+        Returns
+        -------
+            datetime.datetime object
+        """
+        if self.time_units == "unknown":
+            raise AssertionError(
+                "time units must be set in order to calculate datetime"
+            )
+
+        totim = self.get_totim(kper=kper, kstp=kstp)
+
+        if self.time_units == "years":
+            ndays = 365
+            years = np.floor(totim)
+            year = self.start_datetime.year + years
+            if self.start_datetime.month > 2:
+                isleap = calendar.isleap(year + 1)
+            else:
+                isleap = calendar.isleap(year)
+
+            if isleap:
+                ndays = 366
+
+            days = ndays * (totim - years)
+            day_td = datetime.timedelta(days=days)
+
+            dt = datetime.datetime(
+                year,
+                self.start_datetime.month,
+                self.start_datetime.day,
+                self.start_datetime.hour,
+                self.start_datetime.minute,
+                self.start_datetime.second,
+            )
+
+            dt += day_td
+
+        else:
+            kwargs = {self.time_units: totim}
+            dt = self.start_datetime + datetime.timedelta(**kwargs)
+
+        return dt
+
+    def intersect(self, datetime_obj=None, totim=None, kper_kstp=False, forgrive=False):
```

Comment:
In MF6 there is a type called `TimeSelect` which manages timestamps with reference to tdis. It can slice them by stress period / time step which is like the inverse of this method. To achieve something similar i.e. map kper/kstp to a time window in either simulation time `(totim, totim + delt)` or real time `(start, end)` I guess one could call `get_totim()` and `get_datetime()` twice.. maybe a `span()` or overloaded `intersect()` or similar could be convenient for this. Just a thought, not sure if it has any uses cases.

Reply:
I'm having a difficult time imagining what this looks like, maybe we can chat more about what TimeSelect does/returns